### ISSUE TITLE
chat-with-rendezvous connection error

### ISSUE THREAD

**Issue by BaldEagleX02** [libp2p/go-libp2p-examples#70]
Hi,
I fired up a simple-bootstrap-node on a server of mine.
Then a friend of mine in another city and I compiled chat-with-rendezvous as described and ran it (./chat -peer <string/from/bootstrap/node> -rendezvous testing).
Our consoles connect successfully to the bootstrap-node and print "Announcing ourselves" and then we get either nothing (literally nothing) or some warnings (they find each other but can't connect).

**Comment by BaldEagleX02** [comment]
@lbordino73 (the friend)

**Comment by BaldEagleX02** [comment]
@lbordino (the friend)

**Comment by BaldEagleX02** [comment]
@bigs Any suggestions?

**Comment by bigs** [comment]
@BaldEagleX02 Are either of you behind NAT? If so, you won't be able to get a connection between the two of you. I'd recommend passing in an explicit listen address, enabling port-forwarding in your router, and trying again. Keep in mind that DHT-based discovery can take some time to kick in. You'll know you've successfully found each other when you see a print message from [this line](https://github.com/libp2p/go-libp2p-examples/blob/master/chat-with-rendezvous/chat.go#L183).

**Comment by BaldEagleX02** [comment]
I modified the log level to debug and then I saw that we discovered each other (after the program printed "Searching for peers..." I saw "Found another peer" or something similar).
We aren't behind NAT as we have static public IPs from our provider.
We tried with ports opened but it didn't work.
We haven't tried with the listen option yet.

**Comment by BaldEagleX02** [comment]
I am hoping that this library works with NAT too, as it handles hole-punching and UPnP.

**Comment by bigs** [comment]
Given the configuration of that specific example, you'll have to set the listen option to reflect your public IP and an accessible port. Another approach would be to enable port mapping via the [NATPortMap](https://github.com/libp2p/go-libp2p/blob/d87f89314c795f87f32b0c900243728f1481ddb7/options.go#L264) option in the call to `libp2p.New`.

**Comment by BaldEagleX02** [comment]
Yeah, in fact this example works very well with UPnP but if I am behind NAT I don't know any accessible port. I have to use hole punching to know at least one and I was wondering if it will be implemented in this library. Obviously it would require a server. Take a look at this example: https://github.com/wilfreddenton/udp-hole-punching
It seems very well structured and, as its name suggests, implements hole punching.
